pkg/server/routerfactory.go

```diff
@@ -31,6 +31,8 @@ type RouterFactory struct {
 
 	chainBuilder *middleware.ChainBuilder
 	tlsManager   *tls.Manager
+
+	cancelPrevState func()
```

Comment:
I would use type `context.CancelFunc` for self-documenting purposes.

Reply:
I have just copied this over from v3 branch, to simplify future merges if necessary.